File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第6章修改数组.cpp
//修改数组 2019年第十届省赛  lanqiaoOJ 185
#include<bits/stdc++.h>
using namespace std;
const int N=1000002;
int A[N];
int s[N];  //并查集
int find_set(int x){            //用“路径压缩”优化的查询
    if(x != s[x])  s[x] = find_set(s[x]);   //路径压缩
    return s[x];
}
int main(){
    for(int i=1;i<N;i++)  s[i]=i; //并查集初始化
    int n;   scanf("%d",&n);
    for(int i=1;i<=n;i++){
        scanf("%d",&A[i]);
        int root = find_set(A[i]);    //查询到并查集的根
        A[i] = root;
        s[root] = find_set(root+1);   //加1
    }
    for(int i=1;i<=n;i++)  printf("%d ",A[i]);
    return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第6章修改数组.py
#修改数组 2019年第十届省赛  lanqiaoOJ 185
def find_set(x):       #有路径压缩优化的查询
    if(x != s[x]):   s[x] = find_set(s[x])     #集的根是最新的一个数
    return s[x]
N=1000002 
s = list(range(N))     #并查集，定义、初始化 s=[0,1,2,3,……]
n = int(input())
A=[int(i) for i in input().split()]
for i in range(n):
    root = find_set(A[i])
    A[i] = root
    s[root] = find_set(root+1)     #加1
for i in A:  print(i,end = ' ')
